File: apps/api-go/internal/aiql/aiql.go
package aiql

import (
	"encoding/json"
	"fmt"
)

// Schema represents the database schema returned by the Rust core.
type Schema struct {
	Tables map[string]Table `json:"tables"`
}

type Table struct {
	Name        string       `json:"name"`
	Columns     []Column     `json:"columns"`
	Indexes     []Index      `json:"indexes"`
	ForeignKeys []ForeignKey `json:"foreign_keys"`
	Description *string      `json:"description"`
}

type Column struct {
	Name         string  `json:"name"`
	DataType     string  `json:"data_type"`
	IsNullable   bool    `json:"is_nullable"`
	IsPrimaryKey bool    `json:"is_primary_key"`
	DefaultValue *string `json:"default_value"`
	Description  *string `json:"description"`
}

type Index struct {
	Name     string   `json:"name"`
	Columns  []string `json:"columns"`
	IsUnique bool     `json:"is_unique"`
}

type ForeignKey struct {
	ConstraintName string `json:"constraint_name"`
	ColumnName     string `json:"column_name"`
	ForeignTable   string `json:"foreign_table"`
	ForeignColumn  string `json:"foreign_column"`
}

// GetSchema fetches and parses the database schema.
func GetSchema(dbURL string) (*Schema, error) {
	jsonStr, err := CrawlPostgres(dbURL)
	if err != nil {
		return nil, fmt.Errorf("aiql core error: %w", err)
	}

	var schema Schema
	if err := json.Unmarshal([]byte(jsonStr), &schema); err != nil {
		return nil, fmt.Errorf("failed to parse schema JSON: %w", err)
	}

	return &schema, nil
}

File: apps/web/lib/aiql.ts
import { api } from "./api";

export interface Column {
  name: string;
  data_type: string;
  is_nullable: boolean;
  is_primary_key: boolean;
  default_value: string | null;
  description: string | null;
}

export interface ForeignKey {
  constraint_name: string;
  column_name: string;
  foreign_table: string;
  foreign_column: string;
}

export interface Table {
  name: string;
  columns: Column[];
  foreign_keys: ForeignKey[];
  indexes: any[];
}

export interface Schema {
  version: string;
  created_at: string;
  tables: Record<string, Table>;
}

export interface TranslateResult {
  type: "plan" | "clarification";
  dialect?: string;
  raw_query?: string;
  explanation?: string;
  reason?: string;
  suggestions?: string[];
}

export const aiqlApi = {
  crawl: async (url: string): Promise<Schema> => {
    const { data } = await api.post("/api/aiql/crawl", { url });
    return data;
  },
  translate: async (prompt: string, schema: Schema): Promise<TranslateResult> => {
    const { data } = await api.post("/api/aiql/translate", { 
      prompt, 
      schema_json: JSON.stringify(schema) 
    });
    return data;
  },
  export: async (name: string, path: string, prompt: string, sql: string): Promise<any> => {
    const { data } = await api.post("/api/endpoints", {
      name,
      path,
      prompt,
      sql,
      method: "GET"
    });
    return data;
  }
};
